### PR TITLE
server: add HSTS support via cluster setting

### LINKED ISSUE #77224: server: support HSTS header for HTTP connections on secure clusters
**Is your feature request related to a problem? Please describe.**
When a cluster is running in secure mode, an operator should be able to enable HSTS headers on requests to ensure that browsers only connect to the DB Console using HSTS.

**Describe the solution you'd like**
This feature should be off by default (to allow for operator-driven configuration of certs and DNS before enforcing things) and be enabled via a cluster setting named something like `security.hsts.enabled`. Once this is true, HTTP requests to nodes in the cluster will be automatically upgraded to HTTPS by valid user agents.

From casually thinking this through I don't think the addition of this will break any user agents that *don't* look at this header since they'll just still use the HTTP -> HTTPS redirect. However, enabling this with an invalid HTTPS configuration could result in an unreachable DB Console from a web browser so we should take care to document properly.

**Additional context**
* HSTS Explanation https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security
* HTTP redirects are strictly less secure than HSTS because the initial HTTP request could be MITMed to redirect somewhere unsafe. With HSTS *all* requests in the browser after the initial one will use HTTPS and no redirect is necessary.

Jira issue: CRDB-13487

### PR DESCRIPTION
**Pull request by dhartunian**
This commit adds a boolean cluster setting:
`server.hsts.enabled`
which, when enabled attaches standard HSTS headers to http requests
originating from CRDB nodes. The headers looke like this:
`Strict-Transport-Security: max-age=31536000`

When this header is present, most web browsers will automatically
upgrade all HTTP connections to HTTPS and *remember* that setting until
the expiry of 1 year defined in the header.

*Important*: Careless enabling of this feature can result in broken
access to the DB Console in the browser. If we instruct the browser to
always use HTTPS without having a valid TLS configuration, the browser
will no longer fallback to HTTP until the HSTS setting is manually
cleared.

Resolves https://github.com/cockroachdb/cockroach/issues/77224

Release justification: low-risk update; opt-in security enhancement

Release note (security update, ops change, ui change): users can enable
HSTS headers to be set on all HTTP requests which force browsers to
upgrade to HTTPS without a redirect. This is controlled by setting the
`server.hsts.enabled` cluster setting to true which is false by default.

### CODE REVIEW

**Review by abarganier — APPROVED** [review]
:lgtm:, thanks for handling! Just one optional nit re: clarification in the cluster setting's description.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77244#-:-Mx696dc8LY22KGKeq7h:bffdpd7)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian and @knz)

___
*[pkg/server/server_http.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77244#-Mx690i6BoDiiZ2RPxtR:-Mx690i72QLNlJyqDRS3:b-2rz6mg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4b06ff08657e1b2643d5f7c996d8b862d1ebb0e/pkg/server/server_http.go#L66)):*
> ```Go
> 		"security.hsts.enabled",
> 		"if true, HSTS headers will be sent along with all HTTP "+
> 			"requests. This will disable the insecure healthcheck aviailable "+
> ```

nit: should we note the specific `max-age` value in the description?



_Code quote:_
```Go
		"if true, HSTS headers will be sent along with all HTTP "+
			"requests. This will disable the insecure healthcheck aviailable "+
			"at /health.",
```

**Review by abarganier — CHANGES_REQUESTED** [review]
Ah! Also - let's add a test or two :) 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77244#-:-Mx6AKDCCw2RVFd0_Eyz:b-7158zu)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian and @knz)

**Review by zhoufang-joe — COMMENTED** [review]
I was redirected by support engineer from https://support.cockroachlabs.com/hc/en-us/requests/11552. 
Just curious is this the http server used everywhere, including the web server?
I was looking at the https://github.com/cockroachdb/cockroach/blob/9bab39faaf33dfe1f52b3edccc3f758fa6697f40/pkg/ui/ui.go#L120, which is specifically for the UI page for the web browser.

File: pkg/server/server_http.go
Comment (zhoufang-joe) [inline pkg/server/server_http.go:64]: Once this code is checked in, is this config turned via start command or SQL command? Thx

Comment (zhoufang-joe) [reply pkg/server/server_http.go:64]: @dhartunian can you help with the above question ^ ? Thx

Comment (dhartunian) [reply pkg/server/server_http.go:64]: Once this is checked-in, you can enable/disable via SQL:

```
set cluster setting security.hsts.enabled = true;
```

or 

```
set cluster setting security.hsts.enabled = false;
```

Comment (dhartunian) [reply pkg/server/server_http.go:64]: if you would like to use the cluster setting I would advise waiting until we merge this so the naming of the setting is settled. otherwise future upgrades could create regressions.

**Comment by dhartunian** [comment]
> I was redirected by support engineer from https://support.cockroachlabs.com/hc/en-us/requests/11552. Just curious is this the http server used everywhere, including the web server? I was looking at the
> 
> https://github.com/cockroachdb/cockroach/blob/9bab39faaf33dfe1f52b3edccc3f758fa6697f40/pkg/ui/ui.go#L120
> 
> , which is specifically for the UI page for the web browser.

@zhoufang-joe you are correct the the code in `ui.go` would allow you to set the header on just that page. However, I'm not sure that would satisfy your HSTS requirements. That would omit the headers on requests to, say, the prometheus endpoints on `_status/vars` if you're using them.

I'd can change the implementation here if you think it's better to *just* set it on the `index.html` handler but I assumed we needed wider coverage for the header.

File: pkg/server/server_http.go
Comment (zhoufang-joe) [reply pkg/server/server_http.go:64]: Yeah, sure, I will wait for your PR to merge. Then I should be able to pull the latest version of `v21.1` (we are currently at v21.1.12). Thx

Comment (dhartunian) [reply pkg/server/server_http.go:64]: Ah, if you need this on `21.1`, I will need to backport the change. I will set the labels on this PR so that this gets done. 

Once we merge to `master` that will get it into the upcoming `22.1` release. Any feature that needs to be present on `21.1` or `21.2` will need to be backported.

**Comment by zhoufang-joe** [comment]
> > I was redirected by support engineer from https://support.cockroachlabs.com/hc/en-us/requests/11552. Just curious is this the http server used everywhere, including the web server? I was looking at the
> > https://github.com/cockroachdb/cockroach/blob/9bab39faaf33dfe1f52b3edccc3f758fa6697f40/pkg/ui/ui.go#L120
> > 
> > , which is specifically for the UI page for the web browser.
> 
> @zhoufang-joe you are correct the the code in `ui.go` would allow you to set the header on just that page. However, I'm not sure that would satisfy your HSTS requirements. That would omit the headers on requests to, say, the prometheus endpoints on `_status/vars` if you're using them.
> 
> I'd can change the implementation here if you think it's better to _just_ set it on the `index.html` handler but I assumed we needed wider coverage for the header.

I think you are right. It should be available at all UI pages, the prometheus endpoint should have that too.
My intention here is that such header should not be there for the SQL port. Given SQL is not intended for the web browser case anyway. But you probably have better understanding. I trust your judgement. Thx

File: pkg/server/server_http.go
Comment (zhoufang-joe) [reply pkg/server/server_http.go:64]: @dhartunian I see, please do backport to `21.1`. We intentionally keep one version lower than the current GA one, assuming new versions need some time to be mature. When `22.1` becomes GA, we may want to go to `21.2`. But for now, we will use `21.1`. Thx for the quick turnaround!

**Review by knz — APPROVED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77244)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier and @dhartunian)

___
*[pkg/server/server_http.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77244#-MxGxpNH3Ng8lbFZ_LxA:-MxGxpNIDLxw0KN8nllm:b-43exbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6e55503f8864190188ac25f6e973a946b5629d3/pkg/server/server_http.go#L81)):*
> ```Go
> // of 1 year that the browser should remember to only use HTTPS for
> // this site.
> const hstsHeaderValue = "max-age=31536000"
> ```

same question as the other PR - any rationale for chosing 1 year as opposed to any other duration?

**Review by dhartunian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77244)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @dhartunian, and @knz)

___
*[pkg/server/server_http.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77244#-MxGxpNH3Ng8lbFZ_LxA:-Mxg0ekB0CNWYXLFOw9M:bkywtwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6e55503f8864190188ac25f6e973a946b5629d3/pkg/server/server_http.go#L81)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

same question as the other PR - any rationale for chosing 1 year as opposed to any other duration?
</blockquote></details>

The requirements for submitting to the Chrome HSTS preload list stipulate one year as a minimum so it seemed like a sensible value to choose here. https://hstspreload.org/
I don't see much utility in making it user-configurable and it's possible some strict vuln scanners might flag values lower than 1 year.

**Comment by knz** [comment]
thanks for explaining - my review approval remains unchanged.

**Comment by dhartunian** [comment]
Quick update before I merge this (cc @zhoufang-joe) but I've modified the cluster setting to have the `server.` prefix NOT the `security.` prefix. 

My reasoning is that the HSTS headers a property of the server specifically so the prefix is accurate. Plus, their use doesn't directly affect the security of the cluster itself or its subsystems, only the connection between the browser and the server *if* the user agent chooses to follow it so it's more of a piece of guidance to the browser than a security restriction.

**Comment by zhoufang-joe** [comment]
Sure, it sounds good to me. Thanks for the heads up.

**Comment by dhartunian** [comment]
bors r=knz

**Comment by dhartunian** [comment]
@zhoufang-joe this change is now backported to 21.1. Please keep in mind that the support window for 21.1 is closing soon (https://www.cockroachlabs.com/docs/releases/release-support-policy.html#current-supported-releases) but the next release (21.1.17 I believe) will contain this feature.

**Pull request merged**
